File: src/main/java/CardB3.java
import java.util.Objects;

public class CardB3 implements Comparable<CardB3>{
    private int rank;
    private Suit suit;

    public CardB3 (int rank, String suit){
        try {
            this.suit = Suit.valueOf(suit);
        } catch (IllegalArgumentException e){
            throw new IllegalArgumentException("Некорректное значение поля suit");
        }
        this.rank = rank;
    }

    public int getRank() {
        return rank;
    }

    public String getSuit() {
        return suit.name();
    }

    @Override
    public String toString() {
        String stringRank = switch (rank) {
            case 1 -> "Ace";
            case 11 -> "Jack";
            case 12 -> "Queen";
            case 13 -> "King";
            case 14 -> "Joker";
            default -> Integer.toString(rank);
        };
        return this.getSuit() + " " + stringRank;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass() || hashCode() != o.hashCode()) return false;
        CardB3 cardB3 = (CardB3) o;
        return rank == cardB3.rank && suit == cardB3.suit;
    }

    @Override
    public int hashCode() {
        return Objects.hash(rank, suit);
    }

    /* Джокеры в колоде обычно бывают красными и черными и не привязаны к масти.
    Но чтобы не вводить цвета для джокеров, было принято решение заполнятьь поле suit для них
    сильной мастью из черных(clubs) и из красных(diamonds).
     */
    public boolean validate() {
        return (rank > 0 && rank < 14) || (rank == 14 && (suit == Suit.spades || suit == Suit.hearts));
    }

    public boolean isHigher(CardB3 anotherCard) {
        boolean isRedJoker = this.rank == 14 && suit == Suit.hearts;
        boolean isBlackJoker = this.rank == 14 && suit == Suit.spades;
        return ((this.suit.equals(anotherCard.suit) || (isRedJoker && anotherCard.suit == Suit.diamonds)
                || (isBlackJoker && anotherCard.suit == Suit.clubs)) && this.rank > anotherCard.rank);
    }

    /* В задании предполагалось использование массива, что было бы актуально для ситуации,
     когда Suit хранится в строковом варианте. Так как изначально я использовала перечисляемый тип для этого,
     то при выполнении задания я смогла отказаться от массивов, используя индексы элементов перечисляемого типа.
     */
    @Override
    public int compareTo(CardB3 anotherCard){
        return (rank - anotherCard.rank) * 4 + (suit.ordinal() - anotherCard.suit.ordinal());
    }

    public static int compare(CardB3 firstCard, CardB3 secondCard) {
        return firstCard.compareTo(secondCard);
    }

}

File: src/main/java/CardB3.kt
import java.util.*

class CardB3(val rank_value:Int, val suit_value:String) : Comparable<CardB3> {
    private var rank: Int = rank_value
    private var suit: Suit = Suit.valueOf(suit_value)

    fun getRank() : Int {
        return rank
    }

    fun getSuit() : String {
        return suit.name
    }

    override fun toString():String {
        var stringRank: String = when (rank) {
            1 -> "Ace";
            11 -> "Jack";
            12 -> "Queen";
            13 -> "King";
            14 -> "Joker";
            else -> Integer.toString(rank);
        }
        return this.getSuit() + " " + stringRank
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || hashCode() != other.hashCode()) return false
        return when(other) {
            is CardB3 -> rank == other.rank && suit == other.suit;
            else -> false;
        }
    }

    override fun hashCode(): Int {
        return Objects.hash(rank, suit);
    }

    /* Джокеры в колоде обычно бывают красными и черными и не привязаны к масти.
    Но чтобы не вводить цвета для джокеров, было принято решение заполнять поле suit для них
    сильной мастью из черных(clubs) и из красных(diamonds).
     */
    fun validate(): Boolean {
        return (rank in 1..13) || (rank == 14 && (suit == Suit.spades || suit == Suit.hearts));
    }

    fun isHigher(anotherCard:CardB3):Boolean {
        val isRedJoker:Boolean = this.rank == 14 && suit == Suit.hearts;
        val isBlackJoker:Boolean = this.rank == 14 && suit == Suit.spades;
        return ((this.suit == anotherCard.suit || (isRedJoker && anotherCard.suit == Suit.diamonds)
                || (isBlackJoker && anotherCard.suit == Suit.clubs)) && this.rank > anotherCard.rank);
    }

    companion object {
        fun compare(a: CardB3, b: CardB3):Int = a.compareTo(b)
    }

    override fun compareTo(other: CardB3): Int {
        return (rank - other.rank) * 4 + (suit.ordinal - other.suit.ordinal);
    }
}